src/common/btree.c

```diff
@@ -4640,3 +4675,178 @@ dbtree_overhead_get(int alloc_overhead, unsigned int tclass, uint64_t otype,
+
+/**
+ * Validate the integrity of the btree node.
+ *
+ * \param[in] nd	Node to check.
+ * \param[in] nd_off	Node's offset.
+ * \param[in] ck	Checker.
+ *
+ * \retval DER_SUCCESS	The node is correct.
+ * \retval -DER_NOTYPE	The node is malformed.
+ */
+static int
+btr_node_check(struct btr_node *nd, umem_off_t nd_off, struct checker *ck)
+{
+	uint16_t unknown_flags;
+
+	D_ASSERT(ck != NULL);
+
+	CK_PRINTF(ck, "Node (off=%#x)... ", nd_off);
+
+	unknown_flags = nd->tn_flags & ~(BTR_NODE_LEAF | BTR_NODE_ROOT);
+	if (unknown_flags != 0) {
+		CK_APPENDFL_ERR(ck, CK_BTREE_NODE_MALFORMED_STR "unknown flags (%#" PRIx16 ")",
+				unknown_flags);
+		return -DER_NOTYPE;
+	}
+
+	if (nd->tn_pad_32 != 0) {
+		if (ck->ck_options.cko_non_zero_padding == CHECKER_EVENT_ERROR) {
+			CK_APPENDFL_ERR(ck, CK_BTREE_NON_ZERO_PADDING_FMT, nd->tn_pad_32);
+			return -DER_NOTYPE;
+		} else {
+			CK_APPENDFL_WARN(ck, CK_BTREE_NON_ZERO_PADDING_FMT, nd->tn_pad_32);
+		}
+	}
+
+	if (nd->tn_gen != 0) {
+		if (ck->ck_options.cko_non_zero_padding == CHECKER_EVENT_ERROR) {
+			CK_APPENDFL_ERR(ck, CK_BTREE_NON_ZERO_GEN_FMT, nd->tn_gen);
+			return -DER_NOTYPE;
+		} else {
+			CK_APPENDFL_WARN(ck, CK_BTREE_NON_ZERO_GEN_FMT, nd->tn_gen);
+		}
+	}
+
+	CK_APPENDL_OK(ck);
+
+	return DER_SUCCESS;
+}
+
+/**
+ * \struct node_info
+ *
+ * List of node offsets.
+ */
+struct node_info {
+	d_list_t   link;
+	umem_off_t nd_off;
+};
+
+/**
+ * Validate the integrity of a btree.
+ *
+ * \param[in] tcx		Btree context.
+ * \param[in] ck		Checker.
+ *
+ * \retval DER_SUCCESS		The tree is correct.
+ * \retval -DER_NOTYPE		The tree is malformed.
+ * \retval -DER_NONEXIST	The tree is malformed.
+ * \retval -DER_*		Possibly other errors.
+ */
+static int
+btr_nodes_check(struct btr_context *tcx, struct checker *ck)
+{
+	D_LIST_HEAD(node_list);
+	struct node_info *ni;
+	struct node_info *ni_tmp;
+	umem_off_t        nd_off;
+	struct btr_node  *nd;
+	int               rc = DER_SUCCESS;
+
+	D_ASSERT(ck != NULL);
+
+	if (btr_root_empty(tcx)) {
+		CK_PRINT(ck, "Empty tree\n");
+		return DER_SUCCESS;
+	}
+
+	D_ASSERT(!btr_has_embedded_value(tcx));
+
+	/** add the root node to the list */
+	D_ALLOC_PTR(ni);
+	ni->nd_off = tcx->tc_tins.ti_root->tr_node;
+	d_list_add_tail(&ni->link, &node_list);
+
+	/** process the list */
+	d_list_for_each_entry_safe(ni, ni_tmp, &node_list, link) {
+		/** pick the node from the list's head */
+		nd_off = ni->nd_off;
+		nd     = btr_off2ptr(tcx, nd_off);
+
+		/** check the node */
+		rc = btr_node_check(nd, nd_off, ck);
+		if (rc != DER_SUCCESS) {
+			break;
+		}
+
+		/** remove the node from the list */
+		d_list_del(&ni->link);
+		D_FREE(ni);
+
+		/** a leaf has no child nodes */
+		if (nd->tn_flags & BTR_NODE_LEAF) {
+			continue;
+		}
+
+		/** append the node's children to the list */
+		for (int at = 0; at < nd->tn_keyn - 1; ++at) {
+			D_ALLOC_PTR(ni);
+			ni->nd_off = btr_node_child_at(tcx, nd_off, at);
+			d_list_add_tail(&ni->link, &node_list);
```

Comment:
How can this work? I think 'list_for_each_entry_safe()' is only safe against removal. I think current implementation checks only the root node, no?

Reply:
Well spotted. I planned to take care of it when the overall idea gets approvals. But since you pointed it out, you are right. It does not work this way. I wrote a mini-unit test to test it and I made some improvements. Please take a look.

The actual unit test to test it thoroughly is planned for later.